SCons/Tool/ninja.py

```diff
@@ -0,0 +1,1570 @@
+            cmd = cmd[0:-2].strip()
+
+    # Escape dollars as necessary
+    return cmd.replace("$", "$$")
+
+
+def get_shell_command(env, node, action, targets, sources, executor=None):
+    return (
+        "GENERATED_CMD",
+        {
+            "cmd": generate_command(env, node, action, targets, sources, executor=None),
+            "env": get_command_env(env),
+        },
+    )
+
+
+def get_command(env, node, action):  # pylint: disable=too-many-branches
+    """Get the command to execute for node."""
+    if node.env:
+        sub_env = node.env
+    else:
+        sub_env = env
+
+    executor = node.get_executor()
+    if executor is not None:
+        tlist = executor.get_all_targets()
+        slist = executor.get_all_sources()
+    else:
+        if hasattr(node, "target_peers"):
+            tlist = node.target_peers
+        else:
+            tlist = [node]
+        slist = node.sources
+
+    # Retrieve the repository file for all sources
+    slist = [rfile(s) for s in slist]
+
+    # Generate a real CommandAction
+    if isinstance(action, SCons.Action.CommandGeneratorAction):
+        # pylint: disable=protected-access
+        action = action._generate(tlist, slist, sub_env, 1, executor=executor)
+
+    variables = {}
+
+    comstr = get_comstr(sub_env, action, tlist, slist)
+    if not comstr:
+        return None
+    
+    provider = __NINJA_RULE_MAPPING.get(comstr, get_shell_command)
+    rule, variables = provider(sub_env, node, action, tlist, slist, executor=executor)
+
+    # Get the dependencies for all targets
+    implicit = list({dep for tgt in tlist for dep in get_dependencies(tgt)})
+
+    ninja_build = {
+        "order_only": get_order_only(node),
+        "outputs": get_outputs(node),
+        "inputs": get_inputs(node),
+        "implicit": implicit,
+        "rule": rule,
+        "variables": variables,
+    }
+
+    # Don't use sub_env here because we require that NINJA_POOL be set
+    # on a per-builder call basis to prevent accidental strange
+    # behavior like env['NINJA_POOL'] = 'console' and sub_env can be
+    # the global Environment object if node.env is None.
+    # Example:
+    #
+    # Allowed:
+    #
+    #     env.Command("ls", NINJA_POOL="ls_pool")
+    #
+    # Not allowed and ignored:
+    #
+    #     env["NINJA_POOL"] = "ls_pool"
+    #     env.Command("ls")
+    #
+    if node.env and node.env.get("NINJA_POOL", None) is not None:
+        ninja_build["pool"] = node.env["NINJA_POOL"]
+
+    return ninja_build
+
+
+def ninja_builder(env, target, source):
+    """Generate a build.ninja for source."""
+    if not isinstance(source, list):
+        source = [source]
+    if not isinstance(target, list):
+        target = [target]
+
+    # We have no COMSTR equivalent so print that we're generating
+    # here.
+    print("Generating:", str(target[0]))
+
+    generated_build_ninja = target[0].get_abspath()
+    NINJA_STATE.generate()
+
+    if env["PLATFORM"] == "win32":
```

Comment:
As an end user, I found it a bit annoying that I had to always execute ninja after my scons command, so I put at as an extra step after scons is done generating the ninja file. I added a tool option to disable auto execution.

I also found that on windows, Ninja does not have a great way to take in an alternate env (Linux works with prepending environment variables to the command) so this would also run ninja in a subprocess with the correct environment as was setup with the scons build.

Reply:
I think you should consider not doing this:

- Yes, you need to re-invoke Ninja. But that is the right thing to do. You can't know what targets the user actually wants to build. The only target you can name to SCons when building is the ninja file itself, so you have no information about what they actually want to build. The default target might be expensive to build, and maybe they just wanted a particular unit test.

- Regarding the windows environment - yeah, that is a thing. But if you wanted a hermetic build, you should have stayed with SCons driving the build. Once you engage with Ninja, you are building a snapshot and you are at the mercy of the local environment. We actually do manage to pass environment variables down in the generated Ninja on Windows, btw.

Reply:
> The only target you can name to SCons when building is the ninja file itself, so you have no information about what they actually want to build.

I found this a bit confusing that you had to name ninja-generation as a target to generate ninja for the mongodb build. I agree that the ninja file is an output of the build, but I was approaching the ninja file not like a real target. In my opinion its more like an extra build step to get to the actual output of the build. I don't see why targets provided to SCons could not be passed down to ninja command? SCons should also be able to tell if the ninja file needs to be regenerated and just skip to ninja execution with the new targets if not. Anyways, this was more of a personal UIX thing, so its not really functional. I am completely fine with not doing it. 

> But if you wanted a hermetic build, you should have stayed with SCons driving the build.

What if I want a hermetic build and ninja speed? I don't think its impossible, maybe it should be deferred and voted on, but I don't think that's unreasonable want.

> We actually do manage to pass environment variables down in the generated Ninja on Windows, btw.

Any tips on where and what that looks like? I am trying to setup the tool so that a new user to SCons can almost effortless get a SCons with ninja build working. Needing to preconfigure a build environment is not a pro, but if there is a way for the build to do this for the user that is a pro.